File: Modules/Learn/Sources/Learn/Matn.swift
import Foundation

/// A classical didactic poem (متن) the user memorises, as bundled by
/// `Tools/build_matn_tuhfa.py`.
///
/// The schema is deliberately wider than today's needs: it carries N matns
/// (al-Jazariyyah drops straight in once a *vowelled* source is verified —
/// the Wikisource copy is unvowelled, which is useless for tajweed), and it
/// reserves `audio`/`timings` so follow-along recitation can be added without
/// restructuring anything. No audio ships until a recording's licence is
/// recorded in LICENSES.md (CLAUDE.md rule 5).
public struct Matn: Codable, Identifiable, Hashable, Sendable {
    public let id: String
    public let titleAr: String
    public let titleEn: String
    /// Short citation forms, for the navigation bar — the full title is the
    /// heading inside the reader. Optional so a matn may omit them.
    public let shortTitleAr: String?
    public let shortTitleEn: String?
    public let authorAr: String
    public let authorEn: String
    /// Arabic composition note, e.g. "فرغ من نظمها سنة ١١٩٨ هـ".
    public let composedAr: String?
    public let composedYearHijri: Int?
    public let sourceName: String
    public let sourceUrl: String
    public let sourceLicense: String
    public let retrieved: String
    /// Reserved: a recitation of the whole matn (file name only).
    public let audio: String?
    /// Reserved: per-line start seconds for follow-along highlighting.
    public let timings: [Double]?
    public let sections: [Section]
    public let lines: [Line]

    public struct Section: Codable, Identifiable, Hashable, Sendable {
        public let id: String
        public let titleAr: String
        public let titleEn: String?

        enum CodingKeys: String, CodingKey {
            case id
            case titleAr = "title_ar"
            case titleEn = "title_en"
        }

        public func displayTitle(arabicUI: Bool) -> String {
            arabicUI ? titleAr : (titleEn ?? titleAr)
        }
    }

    /// One بيت: two hemistichs (صدر / عجز).
    public struct Line: Codable, Identifiable, Hashable, Sendable {
        public let number: Int
        public let sectionId: String
        public let first: String
        public let second: String
        /// Reserved: start second of this line in `Matn.audio`.
        public let start: Double?

        public var id: Int { number }

        enum CodingKeys: String, CodingKey {
            case number, first, second, start
            case sectionId = "section_id"
        }
    }

    enum CodingKeys: String, CodingKey {
        case id, sections, lines, audio, timings, retrieved
        case titleAr = "title_ar"
        case titleEn = "title_en"
        case shortTitleAr = "short_title_ar"
        case shortTitleEn = "short_title_en"
        case authorAr = "author_ar"
        case authorEn = "author_en"
        case composedAr = "composed_ar"
        case composedYearHijri = "composed_year_hijri"
        case sourceName = "source_name"
        case sourceUrl = "source_url"
        case sourceLicense = "source_license"
    }

    public func displayTitle(arabicUI: Bool) -> String { arabicUI ? titleAr : titleEn }

    /// The name for a 44pt inline navigation bar: short enough to sit beside
    /// the toolbar buttons without truncating. Falls back to the full title
    /// when a matn declares no short form.
    public func navigationTitle(arabicUI: Bool) -> String {
        (arabicUI ? shortTitleAr : shortTitleEn) ?? displayTitle(arabicUI: arabicUI)
    }
    public func displayAuthor(arabicUI: Bool) -> String { arabicUI ? authorAr : authorEn }

    /// The lines of one section, in order.
    public func lines(in section: Section) -> [Line] {
        lines.filter { $0.sectionId == section.id }
    }
}

public enum MatnStore {
    /// Every bundled matn. Add a file here (and its builder under `Tools/`)
    /// to ship another one — the UI is already N-matn.
    static let bundledFiles = ["matn-tuhfat-al-atfal"]

    public static func load() -> [Matn] {
        let decoder = JSONDecoder()
        return bundledFiles.compactMap { name in
            guard let url = Bundle.module.url(forResource: name, withExtension: "json"),
                  let data = try? Data(contentsOf: url)
            else { return nil }
            return try? decoder.decode(Matn.self, from: data)
        }
    }

    public static func matn(id: String) -> Matn? {
        load().first { $0.id == id }
    }
}

File: android/app/src/main/java/com/engagendy/noor/Matn.kt
package com.engagendy.noor

import android.content.Context
import org.json.JSONObject

/// A classical didactic poem (متن) the user memorises — 1:1 with the iOS
/// `Modules/Learn/Sources/Learn/Matn.swift` model, reading the SAME JSON
/// files (`assets/matn-*.json` are byte-identical copies of the iOS bundle
/// resources built by `Tools/build_matn_*.py`).
///
/// The schema is deliberately wider than today's needs: it carries N matns,
/// and reserves `audio`/`timings` for follow-along recitation. No audio
/// ships until a recording's licence is recorded in LICENSES.md.
data class MatnSection(val id: String, val titleAr: String, val titleEn: String?) {
    fun displayTitle(): String = if (isArabicLocale()) titleAr else (titleEn ?: titleAr)
}

/// One بيت: two hemistichs (صدر / عجز).
data class MatnLine(
    val number: Int,
    val sectionId: String,
    val first: String,
    val second: String,
)

data class Matn(
    val id: String,
    val titleAr: String,
    val titleEn: String,
    val shortTitleAr: String?,
    val shortTitleEn: String?,
    val authorAr: String,
    val authorEn: String,
    val composedAr: String?,
    val sourceName: String,
    val sourceLicense: String,
    /// True when the section headings are OURS, not the source's — the
    /// reader says so under the poem rather than letting an editorial label
    /// pass for the author's (al-Bayquniyyah's source page has none).
    val sectionsEditorial: Boolean,
    val sections: List<MatnSection>,
    val lines: List<MatnLine>,
) {
    fun displayTitle(): String = if (isArabicLocale()) titleAr else titleEn

    /// Short enough for a screen header beside its buttons; falls back to
    /// the full title when a matn declares no short form.
    fun navigationTitle(): String =
        (if (isArabicLocale()) shortTitleAr else shortTitleEn) ?: displayTitle()

    fun displayAuthor(): String = if (isArabicLocale()) authorAr else authorEn

    fun lines(section: MatnSection): List<MatnLine> = lines.filter { it.sectionId == section.id }
}

object MatnStore {
    /// Every bundled matn. Add a file here (and its builder under `Tools/`)
    /// to ship another one — the UI is already N-matn.
    private val bundledFiles = listOf("matn-tuhfat-al-atfal", "matn-bayquniyyah")

    @Volatile
    private var cache: List<Matn>? = null

    /// Reads and parses the bundled JSON. ~13 KB total, but still call it
    /// off the main thread (the screens do, via `produceState`).
    fun load(context: Context): List<Matn> {
        cache?.let { return it }
        val parsed = bundledFiles.mapNotNull { name ->
            runCatching {
                parse(context.assets.open("$name.json").bufferedReader().use { it.readText() })
            }.getOrNull()
        }
        cache = parsed
        return parsed
    }

    fun matn(context: Context, id: String): Matn? = load(context).firstOrNull { it.id == id }

    private fun parse(json: String): Matn {
        val root = JSONObject(json)
        val sections = root.getJSONArray("sections").let { array ->
            (0 until array.length()).map { index ->
                val obj = array.getJSONObject(index)
                MatnSection(
                    id = obj.getString("id"),
                    titleAr = obj.getString("title_ar"),
                    titleEn = obj.optString("title_en").takeIf { it.isNotEmpty() })
            }
        }
        val lines = root.getJSONArray("lines").let { array ->
            (0 until array.length()).map { index ->
                val obj = array.getJSONObject(index)
                MatnLine(
                    number = obj.getInt("number"),
                    sectionId = obj.getString("section_id"),
                    first = obj.getString("first"),
                    second = obj.getString("second"))
            }
        }
        return Matn(
            id = root.getString("id"),
            titleAr = root.getString("title_ar"),
            titleEn = root.getString("title_en"),
            shortTitleAr = root.optString("short_title_ar").takeIf { it.isNotEmpty() },
            shortTitleEn = root.optString("short_title_en").takeIf { it.isNotEmpty() },
            authorAr = root.getString("author_ar"),
            authorEn = root.getString("author_en"),
            composedAr = root.optString("composed_ar").takeIf { it.isNotEmpty() },
            sourceName = root.getString("source_name"),
            sourceLicense = root.getString("source_license"),
            sectionsEditorial = root.optBoolean("sections_editorial", false),
            sections = sections,
            lines = lines)
    }
}

/// Search inside one matn — every hemistich of every line, folded ONCE when
/// the index is built so a keystroke only scans pre-folded text.
///
/// Nothing here re-implements folding or ranking: `SearchText` is the single
/// Arabic matcher in the app (Quran + athkar search), and the matns are
/// vowelled while people type bare letters, so it is exactly what is needed.
class MatnSearchIndex(matn: Matn) {
    /// One line that matched, and the hemistich the match is in — a snippet
    /// is always a window on text that exists in the poem, never a stitched
    /// string.
    data class Hit(val line: MatnLine, val text: String, val tier: Int)

    // Two documents per line (صدر then عجز).
    private val documents: List<Triple<MatnLine, String, String>> =
        matn.lines.flatMap { line ->
            listOf(
                Triple(line, line.first, SearchText.normalizeForSearch(line.first)),
                Triple(line, line.second, SearchText.normalizeForSearch(line.second)))
        }

    val lineCount: Int = matn.lines.size

    /// Matching lines, best tier first and line order inside a tier. A line
    /// whose two hemistichs both match appears once, at its better tier.
    fun search(query: String): List<Hit> {
        val normalized = SearchText.normalizeForSearch(query.trim())
        if (normalized.isEmpty()) return emptyList()
        val best = LinkedHashMap<Int, Hit>()
        for ((line, text, norm) in documents) {
            if (!norm.contains(normalized)) continue
            val tier = SearchText.matchTier(text, normalized)
            val existing = best[line.number]
            if (existing == null || tier < existing.tier) {
                best[line.number] = Hit(line, text, tier)
            }
        }
        return best.values.sortedWith(compareBy({ it.tier }, { it.line.number }))
    }
}
